contracts/staking/DelegationManager.sol

```diff
@@ -0,0 +1,195 @@
+    _;
+  }
+
+  constructor (Registry _registry) ERC721Full("Matic Delegator", "MD") public {
+    registry = _registry;
+  }
+
+  function setToken(IERC20 _token) public onlyOwner {
+    token = _token;
+  }
+
+  function getRewards(uint256 delegatorId) public onlyDelegator(delegatorId) {
+    _getRewards(delegatorId);
+    require(token.transfer(msg.sender, delegators[delegatorId].reward));
+    delegators[delegatorId].reward = 0;
+  }
+
+  function _getRewards(uint256 delegatorId) internal {
+    address validator;
+    StakeManager stakeManager = StakeManager(registry.getStakeManagerAddress());
+    uint256 currentEpoch = stakeManager.currentEpoch();
+    (,,,,,,validator,) = stakeManager.validators(delegators[delegatorId].bondedTo);
+    delegators[delegatorId].reward += ValidatorContract(validator).getRewards(
+      delegatorId,
+      delegators[delegatorId].amount,
+      delegators[delegatorId].delegationStartEpoch,
+      delegators[delegatorId].delegationStopEpoch,
+      currentEpoch);
+    // require(token.transfer(msg.sender, delegators[delegatorId].reward));
+    delegators[delegatorId].delegationStartEpoch = currentEpoch;
+  }
+
+  function bond(uint256 delegatorId, uint256 validatorId) public onlyDelegator(delegatorId) {
+    // require(time/count)
+    Delegator storage delegator = delegators[delegatorId];
+    StakeManager stakeManager = StakeManager(registry.getStakeManagerAddress());
+    uint256 currentEpoch = stakeManager.currentEpoch(); //TODO add 1
+    // for lazy unbonding
+    if (delegator.delegationStopEpoch != 0 && delegator.delegationStopEpoch < currentEpoch ) {
+      _getRewards(delegatorId);
+      delegator.delegationStopEpoch = 0;
+      delegator.delegationStartEpoch = 0;
+    } else {
+      require(delegator.delegationStopEpoch == 0 &&
+        delegator.delegationStartEpoch == 0 &&
+        delegators[delegatorId].bondedTo == 0,
+        "");
+    }
+    address validator;
+    (,,,,,,validator,) = stakeManager.validators(validatorId);
+    ValidatorContract(validator).bond(delegatorId, delegator.amount);
+    delegator.delegationStartEpoch = currentEpoch;
+    delegator.bondedTo = validatorId;
+    stakeManager.updateValidatorState(validatorId, currentEpoch, int(delegator.amount));
+    emit Bonding(delegatorId, validatorId, validator);
+  }
+
+  function unBond(uint256 delegatorId, uint256 index) public onlyDelegator(delegatorId) {
+    address validator;
+    StakeManager stakeManager = StakeManager(registry.getStakeManagerAddress());
+    _getRewards(delegatorId);
+    (,,,,,,validator,) = stakeManager.validators(delegators[delegatorId].bondedTo);
+    ValidatorContract(validator).unBond(delegatorId, index, delegators[delegatorId].amount);
+    stakeManager.updateValidatorState(delegators[delegatorId].bondedTo, stakeManager.currentEpoch(), -int(delegators[delegatorId].amount));
+    emit UnBonding(delegatorId, delegators[delegatorId].bondedTo);
+    delegators[delegatorId].bondedTo = 0;
```

Comment:
Should we also `delegators[delegatorId].delegationStartEpoch = 0;`?

Reply:
yeah fixed it and then read this comment thanks 🤣